clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -2257,9 +2257,10 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       getContext().BuiltinInfo.isConstWithoutErrnoAndExceptions(BuiltinID);
   bool ConstWithoutExceptions =
       getContext().BuiltinInfo.isConstWithoutExceptions(BuiltinID);
-  if (FD->hasAttr<ConstAttr>() ||
-      ((ConstWithoutErrnoAndExceptions || ConstWithoutExceptions) &&
-       (!ConstWithoutErrnoAndExceptions || (!getLangOpts().MathErrno)))) {
+  if ((FD->hasAttr<ConstAttr>() ||
+       ((ConstWithoutErrnoAndExceptions || ConstWithoutExceptions) &&
+        (!ConstWithoutErrnoAndExceptions || (!getLangOpts().MathErrno)))) &&
+      !(getLangOpts().isSYCL() && getTarget().getTriple().isNVPTX())) {
```

Comment:
All FE changes should have an accompanying FE test. It makes it easier to track down issues in the future if one arises. Can you add a FE test for this change. It can just check these builtins are not present in IR with these options and present without.

Reply:
Sure. Test added. Let me know if you think this is sufficient. Note that SPIR-V compilation does use llvm intrinsics (not libdevice) for cmath funcs when in `-ffast-math` mode (since `errno` not used). This isn't exactly a problem for SPIR-V since `llvm.cos.f32` has a SPIR-V lowering for fast mode, namely `approx_cos`. It happens to work but I think the idea of using libdevice is for DPC++ to provide its own definitions for cmath and other CXX stdlib funcs. Let me know if you think SPIR-V compilation should also use libdevice func definitions while in fast math mode as well, instead of relying on openCL driver implementations for say `approx_cos`.

Reply:
>>Let me know if you think SPIR-V compilation should also use libdevice func definitions while in fast math mode as well, instead of relying on openCL driver implementations for say approx_cos.

I'm not familiar enough with this to have a non-naive response one way or the other. Maybe @bader has an input?